pkg/controllers/state/cost.go

```diff
@@ -0,0 +1,370 @@
+		},
+		[]string{
+			metrics.NodePoolLabel,
+		},
+	)
+)
+
+// ClusterCost tracks the cost of compute resources across all NodePools in a cluster.
+// This is an alpha-level component and its API may change without notice.
+//
+// The ClusterCost maintains real-time cost information by:
+// - Tracking NodeClaim additions and removals
+// - Managing instance type offerings and their prices
+// - Calculating aggregate costs per NodePool and cluster-wide
+//
+// All operations are thread-safe through internal locking mechanisms.
+type ClusterCost struct {
+	sync.RWMutex
+	npCostMap map[string]*NodePoolCost // nodepool.Name -> NodePoolCost
+	// nodeClaimMap tracks which NodeClaims are currently being monitored for cost
+	nodeClaimMap map[string]bool
+
+	cloudProvider cloudprovider.CloudProvider
+	client        client.Client
+}
+
+// NodePoolCost represents the cost tracking information for a single NodePool.
+// It maintains the current cost, available instance types, and count of active offerings.
+type NodePoolCost struct {
+	cost                 float64
+	overlayedInstanceMap map[string]*cloudprovider.InstanceType // instance name -> instance
+	// offeringCounts tracks how many instances of each offering type are currently active
+	offeringCounts map[OfferingKey]OfferingCount
+}
+
+// OfferingKey uniquely identifies a specific compute offering by its zone,
+// capacity type (e.g., spot/on-demand), and instance type name.
+type OfferingKey struct {
+	Zone, Capacity, InstanceName string
+}
+
+// OfferingCount tracks the number and cost of instances for a specific offering.
+type OfferingCount struct {
+	Count int
+	Cost  float64 // cost of the offering, not cost * count
+}
+
+// NewClusterCost creates and initializes a new ClusterCost instance for tracking
+// compute costs across the cluster. It requires a cloud provider for accessing
+// instance type and pricing information, and a Kubernetes client for NodePool lookups.
+func NewClusterCost(ctx context.Context, cloudProvider cloudprovider.CloudProvider, client client.Client) *ClusterCost {
+	return &ClusterCost{
+		npCostMap:     make(map[string]*NodePoolCost),
+		nodeClaimMap:  make(map[string]bool),
+		cloudProvider: cloudProvider,
+		client:        client,
+	}
+}
+
+// UpdateOfferings updates the available instance types and their pricing information
+// for a specific NodePool. This method is typically called when NodePool configurations
+// change or when cloud provider pricing information is refreshed.
+//
+// Returns an error if instance type information cannot be updated or if cost
+// recalculation fails.
+func (cc *ClusterCost) UpdateOfferings(ctx context.Context, np *v1.NodePool, instanceTypes []*cloudprovider.InstanceType) error {
+	cc.Lock()
+	defer cc.Unlock()
+	err := cc.internalUpdateOfferings(np, instanceTypes)
+	if err != nil {
+		return fmt.Errorf("failed to update offerings for nodepool %q: %w", np.Name, err)
```

Comment:
The standard we chose for Karpenter was to use commas to delimit errors rather than colons:
```suggestion
		return fmt.Errorf("failed to update offerings for nodepool %q, %w", np.Name, err)
```